File: java-vfs/src/main/java/emg/java/vfs/files/Directory.java
package emg.java.vfs.files;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

import static emg.java.vfs.extensions.ListExtensions.head;
import static emg.java.vfs.extensions.ListExtensions.tail;

public class Directory extends DirEntry {
    public static final String SEPARATOR = "/";
    public static final String ROOT_PATH = "/";

    private final List<DirEntry> contents;

    public Directory(String parentPath, String name, List<DirEntry> contents) {
        super(parentPath, name);
        this.contents = contents;
    }

    public static Directory root() {
        return Directory.empty("", "");
    }

    public static Directory empty(String parentPath, String name) {
        return new Directory(parentPath, name, new ArrayList<>());
    }

    public boolean hasEntry(String name) {
        return findEntry(name) != null;
    }

    public Directory replaceEntry(String entryName, DirEntry newEntry) {
        List<DirEntry> entryContents = this.contents.stream().filter(d -> !d.name.equals(entryName)).collect(Collectors.toList());
        entryContents.add(newEntry);
        return new Directory(parentPath, name, entryContents);
    }

    public DirEntry findEntry(String entryName) {
        return findEntryHelper(entryName, contents);
    }

    private DirEntry findEntryHelper(String name, List<DirEntry> contentList) {
        if (contentList.isEmpty()) {
            return null;
        } else if (head(contentList).name.equals(name)) {
            return head(contentList);
        } else {
            return findEntryHelper(name, tail(contentList));
        }
    }

    public Directory addEntry(DirEntry newEntry) {
        contents.add(newEntry);
        return new Directory(parentPath, name, contents);
    }

    public Directory findDesendant(List<String> path) {
        if (path.isEmpty()) {
            return this;
        } else {
            return findEntry(head(path)).asDirectory().findDesendant(tail(path));
        }
    }

    public Directory findDesendant(String relativePath) {
        if (relativePath.isEmpty()) {
            return this;
        } else {
            return findDesendant(Arrays.asList(relativePath.split(Directory.SEPARATOR)));
        }
    }

    public List<String> getAllFoldersInPath() {
        return Arrays.stream(path().substring(1).split(Directory.SEPARATOR)).filter(d -> !d.isEmpty()).collect(Collectors.toList());
    }

    @Override
    public Directory asDirectory() {
        return this;
    }

    @Override
    public String getType() {
        return "Directory";
    }

    public List<DirEntry> contents() {
        return contents;
    }

    @Override
    public boolean isDirectory() {
        return true;
    }

    public boolean isRoot() {
        return parentPath.isEmpty();
    }

    public Directory removeEntry(String entryName) {
        if (!hasEntry(entryName)) {
            return this;
        } else {
            return new Directory(parentPath, name, contents.stream().filter(x -> !x.name.equals(entryName)).collect(Collectors.toList()));
        }
    }
}

File: scala-vfs/src/main/scala/emg/scala/vfs/files/Directory.scala
package emg.scala.vfs.files

import scala.annotation.tailrec

class Directory(override val parentPath: String, override val name: String, val contents: List[DirEntry])
  extends DirEntry(parentPath, name) {
  def removeEntry(entryName: String): Directory = {
    if (!hasEntry(entryName)) this
    else new Directory(parentPath, name, contents.filter(x => !x.name.equals(entryName)))
  }


  def replaceEntry(entryName: String, newEntry: DirEntry): Directory =
    new Directory(parentPath, name, contents.filter(d => !d.name.equals(entryName)) :+ newEntry)

  def findEntry(entryName: String): DirEntry = {
    @tailrec
    def findEntryHelper(name: String, contentList: List[DirEntry]): DirEntry = {
      if (contentList.isEmpty) null
      else if (contentList.head.name.equals(name)) contentList.head
      else findEntryHelper(name, contentList.tail)
    }

    findEntryHelper(entryName, contents)
  }

  def addEntry(newEntry: DirEntry): Directory = new Directory(parentPath, name, contents :+ newEntry)

  def findDesendant(path: List[String]): Directory =
    if (path.isEmpty) this
    else findEntry(path.head).asDirectory.findDesendant(path.tail)

  def findDesendant(relativePath: String): Directory = {
    if (relativePath.isEmpty) this
    else findDesendant(relativePath.split(Directory.SEPARATOR).toList)
  }

  def getAllFoldersInPath: List[String] =
    path.substring(1).split(Directory.SEPARATOR).toList.filter(d => !d.isEmpty)

  def hasEntry(name: String): Boolean = findEntry(name) != null

  override def asDirectory: Directory = this

  override def getType: String = "Dir"

  def isRoot: Boolean = parentPath.isEmpty

  override def isDirectory: Boolean = true

  override def toString: String = parentPath + name
}

object Directory {
  val SEPARATOR = "/"
  val ROOT_PATH = "/"

  def ROOT: Directory = Directory.empty("", "")

  def empty(parentPath: String, name: String): Directory = {
    new Directory(parentPath, name, List())
  }
}
